atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/GetCandidateCellsForSweepingShim.java

```diff
@@ -0,0 +1,212 @@
+                                .sortedTimestamps(candidate ? timestampArr : null)
+                                .isLatestValueEmpty(latestValEmpty)
+                                .numCellsTsPairsExamined(numExamined.longValue())
+                                .isCandidate(candidate)
+                                .build());
+                    }
+                }
+                return candidateBatch;
+            });
+            Closer closer = createCloserAndRelease(valueResults, tsResults);
+            return ClosableIterators.wrap(candidates, closer);
+        }
+    }
+
+    private static Closer createCloserAndRelease(ReleasableCloseable<?>... closeables) {
+        Closer closer = Closer.create();
+        for (ReleasableCloseable<?> c : closeables) {
+            closer.register(c.get());
+        }
+        for (ReleasableCloseable<?> c : closeables) {
+            c.release();
+        }
+        return closer;
+    }
+
+    private static boolean isCandidate(long[] timestamps,
+                                       boolean lastValEmpty,
+                                       CandidateCellForSweepingRequest request) {
+        return timestamps.length > 1
+            || (request.shouldCheckIfLatestValueIsEmpty() && lastValEmpty)
+            || (timestamps.length == 1 && timestampIsPotentiallySweepable(timestamps[0], request));
+    }
+
+    private static boolean timestampIsPotentiallySweepable(long ts, CandidateCellForSweepingRequest request) {
+        return ts == Value.INVALID_VALUE_TIMESTAMP || ts >= request.minUncommittedStartTimestamp();
+    }
+
+    private ClosableIterator<RowResult<Value>> getValues(TableReference tableRef,
+                                                         RangeRequest range,
+                                                         long sweepTs,
+                                                         boolean checkIfLatestValueIsEmpty) {
+        if (checkIfLatestValueIsEmpty) {
+            return keyValueService.getRange(tableRef, range, sweepTs);
+        } else {
+            return ClosableIterators.emptyImmutableClosableIterator();
+        }
+    }
+
+    private static boolean isLatestValueEmpty(Cell cell, PeekingIterator<RowResult<Value>> values) {
+        while (values.hasNext()) {
+            RowResult<Value> result = values.peek();
+            int comparison = UnsignedBytes.lexicographicalComparator().compare(cell.getRowName(), result.getRowName());
+            if (comparison == 0) {
+                Value matchingValue = result.getColumns().get(cell.getColumnName());
+                return matchingValue != null && matchingValue.getContents().length == 0;
+            } else if (comparison < 0) {
+                return false;
+            } else {
+                values.next();
+            }
+        }
+        return false;
+    }
+
+    private static class ReleasableCloseable<T extends Closeable> implements AutoCloseable {
```

Comment:
if T extends AutoCloseable instead you could simplify the close method a little (or alternatively make the close() throw Exception to avoid an extra wrapping into a RuntimeException)